crates/oxc_transformer/src/es2018/object_spread.rs

```diff
@@ -0,0 +1,139 @@
+//! ES2018 object spread transformation.
+//!
+//! This plugin transforms object spread properties (`{ ...x }`) to a series of `_objectSpread` calls.
+//!
+//! > This plugin is included in `preset-env`, in ES2018
+//!
+//! ## Example
+//!
+//! Input:
+//! ```js
+//! var x = { a: 1, b: 2 };
+//! var y = { ...x, c: 3 };
+//! ```
+//!
+//! Output:
+//! ```js
+//! var x = { a: 1, b: 2 };
+//! var y = _objectSpread({}, x, { c: 3 });
+//! ```
+//!
+//! ## Implementation
+//!
+//! Implementation based on [@babel/plugin-transform-object-rest-spread](https://babeljs.io/docs/babel-plugin-transform-object-rest-spread).
+//!
+//! ## References:
+//! * Babel plugin implementation: <https://github.com/babel/babel/tree/main/packages/babel-plugin-transform-object-rest-spread>
+//! * Object rest/spread TC39 proposal: <https://github.com/tc39/proposal-object-rest-spread>
+
+use super::object_rest_spread::ObjectRestSpreadOptions;
+use crate::context::Ctx;
+
+use oxc_ast::ast::*;
+use oxc_semantic::ReferenceFlags;
+use oxc_span::SPAN;
+use oxc_traverse::{Traverse, TraverseCtx};
+
+pub struct ObjectSpread<'a> {
+    _ctx: Ctx<'a>,
+    options: ObjectRestSpreadOptions,
+}
+
+impl<'a> ObjectSpread<'a> {
+    pub fn new(options: ObjectRestSpreadOptions, ctx: Ctx<'a>) -> Self {
+        Self { _ctx: ctx, options }
+    }
+}
+impl<'a> Traverse<'a> for ObjectSpread<'a> {
+    fn enter_expression(&mut self, expr: &mut Expression<'a>, ctx: &mut TraverseCtx<'a>) {
+        let Expression::ObjectExpression(obj_expr) = expr else {
+            return;
+        };
+
+        if obj_expr
+            .properties
+            .iter()
+            .all(|prop| matches!(prop, ObjectPropertyKind::ObjectProperty(..)))
+        {
+            return;
+        }
+
+        // collect `y` and `z` from `{ ...x, y, z }`
+        let mut obj_prop_list = ctx.ast.vec();
+        while obj_expr
+            .properties
+            .last()
+            .map_or(false, |prop| matches!(prop, ObjectPropertyKind::ObjectProperty(..)))
+        {
+            let prop = obj_expr.properties.pop().unwrap();
+            obj_prop_list.push(prop);
+        }
+
+        let Some(ObjectPropertyKind::SpreadProperty(mut spread_prop)) = obj_expr.properties.pop()
+        else {
+            unreachable!();
+        };
+
+        self.enter_expression(expr, ctx);
```

Comment:
Here, a pre-order traversal is expected. But I think the order might not matter. I'll test if simply removing it leads to the correct result.
This might be the actual cause of the performance degradation.

Reply:
OK cool. I doubt this is where perf regression is coming from as this code is only reached if a `SpreadElement` is found, and the vast majority of object literals won't use spread.

As far as I can see, going in forwards order should work equally well (but I could of course be wrong). But I tend not to assume that how Babel does things is necessarily the "best way", as Babel has less emphasis on performance than Oxc does, and so doesn't always choose the most efficient algorithms. Our modus operandi tends to be: start by copying Babel, and then try to improve on it!

It's not essential to turn it into a loop, it's just preferable where it's possible, as loops tend to be faster.